crates/winch/src/visitor.rs

```diff
@@ -0,0 +1,713 @@
+use crate::codegen::CodeGen;
+use crate::masm::{MacroAssembler, OperandSize, RegImm};
+use crate::stack::Val;
+use anyhow::Result;
+use wasmparser::{FuncValidator, Result as ValidationResult, ValidatorResources, VisitOperator};
+use wasmtime_environ::WasmType;
+
+/// A macro iterator, inspired by wasmparsers `for_each_operator`,
+/// containing the definition of supported operators
+macro_rules! for_each_supported_op {
+    ($def:ident) => {
+        $def! {
+            @mvp I32Const { value: i32 } => (visit_i32_const emit_i32_const)
+            @mvp I32Add => (visit_i32_add emit_i32_add)
+        @mvp LocalGet { local_index: u32 } => (visit_local_get emit_local_get)
+        @mvp LocalSet { local_index: u32 } => (visit_local_set emit_local_set)
+        }
+    };
+}
+
+/// A macro iterator, inspired by wasmparsers `for_each_operator`,
+/// containing the definition of unsupported operators
+macro_rules! for_each_unsupported_op {
```

Comment:
I'm wondering if it's possible to avoid the complexity of these macros -- I don't like them for several reasons:

* It duplicates knowledge here (the whole set of Wasm opcodes) unnecessarily;
* It's very unreadable;
* The dispatching mechanism feels somewhat "magic" as a result, in a bad way (I can't trace what's going on).

I think I'd prefer a toplevel translation driver loop that's something like:

```rust
for op in operators {
  match op {
    Operator::Add => self.visit_add();
    ...
  }
}
```

maybe with that `match` in a handwritten `fn visit()`. In other words, just expand out what this macro would generate by hand.

Or said another way: it seems these macros are only used once. Given that, why not expand them in the place that they're used, to reduce complexity?

Reply:
> I'm wondering if it's possible to avoid the complexity of these macros

Yeah, definitely possible. I've addressed this in https://github.com/bytecodealliance/wasmtime/pull/4907/commits/147f901ea630bb39b2a59fa956293c6c232f4e41

FWIW, I like the visitor pattern recently introduced in `wasmparser`,  but when I was exploring introducing it I couldn't find a cleaner way to manage the support/unsupported operator sets, which led to duplication. I'd still like to keep the implementation of the `visit` function and each of the emitters in the `visit.rs` module though; I expect that this file is going to get very big as more operators are added.

Reply:
Thanks for chiming in here @alexcrichton; I wasn't aware of the speed benefits; I wrongly assumed the change was mostly around ergonomics. But taking a closer look at the details and [some of the comments](https://github.com/bytecodealliance/wasm-tools/pull/697#issuecomment-1210517471), your comment makes sense to me. 

I'll try out your suggestion, hopefully it'll be less complex that my previous attempt.

Reply:
I do realize that macros are nontrivial to work with for sure though and I'm still not 100% satisfied with where `wasmparser` ended up. If you'd like some help I'm happy to assist, or additionally at this point getting something working is probably more prudent than the highest levels of optimization, so I'm happy to try to come work on this small part as a follow-up if you'd prefer too

Reply:
> Given that winch's goal is compilation speed I might caution against going this direction since the Operator-based approach to both validation and translation is known to be double-digit percentage slower when at least purely validating and would likely have similar impact to the design of winch as well.

Ah, OK, I think we're slightly talking past each other, and my hand expansion into a `match` was based on my misunderstanding of what this was generating (sorry!). My main point though wasn't about the actual matching/control-flow strategy, but about the way we generate it.

Basically my point can be reduced to: there are some macros here; each macro is invoked once; can we inline the macros' expansion at that point instead? (Since this doesn't change what eventually gets compiled into the winch crate, it should be completely performance-neutral.) It didn't seem like the redundancy that the macros reduced (method args, ..) was that significant compared to the complexity of having the macros at all. And my confusion on what they were actually doing is only slightly more evidence of (either my denseness or...) the confusion imparted by macro use :-)

Reply:
> Basically my point can be reduced to: there are some macros here; each macro is invoked once; can we inline the macros' expansion at that point instead? 

I think that in this case, the approach would be to manually define each method expected by the visitor trait, and error out on the ones that are not supported. This satisfies the readability and the performance aspects, at the cost of having to manually define a method per operator.  I'd would still like to explore if there's a middle ground in which we are able to use macros without hindering too much the readability, (e.g. following Alex's suggestion of only having a single list of unsupported operators) and perhaps the complexity can be reduced further with comments? And, thinking a bit more, if we go with this approach, as we support more and more operators, this "unsupported" list is going to shrink over time.

Reply:
Had a chat with Alex (thanks for the help Alex!) in which we were able to simplify the macro. @cfallin let me know if you have further thoughts on this; I'd like to keep a macro if possible, assuming that it will temporary as more operators are supported. If not I'm happy to manually define all the Visitor traits in a follow up PR. In general I feel that this is cleaner than my previous attempt.

Reply:
OK, yeah, I'm happy with how this looks -- thank you for iterating on this!